lib/ansible/cli/console.py

```diff
@@ -558,6 +558,10 @@ def run(self):
 
         hosts = self.get_host_list(self.inventory, context.CLIARGS['subset'], self.pattern)
 
+        # flush fact cache if requested
+        if context.CLIARGS['flush_cache']:
+            self._flush_cache(self.inventory, self.variable_manager)
+
```

Comment:
I think it might be better to do this in the CLI base class `_play_prereqs` method instead (then you can also delete the pattern from playbook.py).

Right now adhoc does flush the cache for inventory plugins, but not the fact cache (which would be fixed by this logic happening in the base class instead). After the setup task, try running `ansible localhost --flush-cache -m debug -a "msg={{ ansible_facts }}"` to see that the fact cache isn't getting cleared yet. It would also be great to add tests for this (maybe in test/integration/targets/adhoc).

Reply:
Moving this logic to the CLI base class seems to work. After running setup task, the fact cache is getting cleared as seen below:
`ansible localhost --flush-cache -m debug -a "msg={{ ansible_facts }}"

<output>
localhost | SUCCESS => {
    "msg": {}
}
`

However  _flush_cache needs to handle "localhost" separately since the hosts_list is empty. Testing further...

Reply:
Good point. It's probably simpler to fix in adhoc after the host list is retrieved.

Reply:
Would it be okay to add something like below to always handle `localhost` whenever `--flush-cache` is used.

```
@staticmethod
    def _flush_cache(inventory, variable_manager):
        **variable_manager.clear_facts('localhost')**
        for host in inventory.list_hosts():
            hostname = host.get_name()
            variable_manager.clear_facts(hostname)
```

Reply:
Yes, adding `variable_manager.clear_facts('localhost')` seems like a good solution.